### PR TITLE
820 duplicate record problem in hs labels

### PR DESCRIPTION
**Pull request by alvacouch**
Several updates to hs_labels: 
* enclosed get_or_create in transaction blocks to eliminate read/create race. 
* added debugging assertions that are only checked when in debug mode. 
* added helper function for flag setting. 

Note that this includes a data migration to remove existing duplicate label and flag records from UserResourceLabels and UserResourceFlags before asserting uniqueness of labels at the model level.

### CODE REVIEW

**Comment by hydrobuild** [comment]
Can one of the admins verify this patch?

**Comment by mjstealey** [comment]
@alvacouch - Looks like an [unhappy test](http://ci.hydroshare.org:8080/job/pull-request-builder/295/testReport/) in the PR build

Error Message
```
cannot import name UserResourceFlags2
```

Stack trace
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/usr/local/lib/python2.7/dist-packages/nose/loader.py", line 414, in loadTestsFromName
    addr.filename, addr.module)
  File "/usr/local/lib/python2.7/dist-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/local/lib/python2.7/dist-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/docker/hydroshare/hs_labels/tests/__init__.py", line 2, in <module>
    from .test_resource_labelling import *
  File "/home/docker/hydroshare/hs_labels/tests/test_resource_labelling.py", line 8, in <module>
    from hs_labels.models import UserLabels, ResourceLabels, \
ImportError: cannot import name UserResourceFlags2
```

**Comment by alvacouch** [comment]
Stupid typo in test suite fixed.

**Comment by alvacouch** [comment]
Oops. I see new failure.... working on it.....

File: docker-compose.yml [hunk 1]
```diff
@@ -25,12 +25,11 @@ hydroshare:
     SOLR_HOST: solr
   volumes:
     # hydroshare repository
-    - ".:/home/docker/hydroshare"
-    #- "/home/hydro/hydroshare:/home/docker/hydroshare"
+    - "/home/hydro/hydroshare:/home/docker/hydroshare"
```
Comment (mjstealey) [inline docker-compose.yml]: No need to include this file in the PR.

The `docker-compose.yml` file gets generated at runtime from `docker-compose.template` using the definitions from `config/hydroshare-config.yaml`.

File: hs_labels/models.py [hunk 2]
```diff
@@ -4,93 +4,106 @@
...
    removes a favorite
 
  and the reporting functions
 
- - u.ulabels.labeled_resources
+ * u.ulabels.labeled_resources
    A queryset of resources that are labeled.
 
- - u.ulabels.favorite_resources
+ * u.ulabels.favorite_resources (trivial, not yet implemented, available tomorrow)
```
Comment (pkdash) [inline hs_labels/models.py]: The function/property name seems to be ```favorited_resources``` and it has been implemented.

File: hs_labels/models.py [hunk 3]
```diff
@@ -4,93 +4,106 @@
...
     FAVORITE = 1
     MINE = 2
     FLAG_CHOICES = (
-        (FAVORITE, 'Favorite'),
-        (MINE, 'Mine'),
+        (FAVORITE, 'Favorite'), # marked as favorite in my resources page.
+        (MINE, 'Mine'),         # marked as mine in discovery page.
     )
 
+# class BaseResource(models.Model):
```
Comment (pkdash) [inline hs_labels/models.py]: I would remove these commented code.

File: hs_labels/models.py [hunk 4]
```diff
@@ -99,25 +112,37 @@ class UserResourceFlags(models.Model):
...
+                             on_delete=models.CASCADE)
+
     resource = models.ForeignKey(BaseResource, null=False, editable=False,
                                  related_name="r2urf",  # unused but must be defined and unique
-                                 help_text='resource to which a flag applies')
+                                 help_text='resource to which a flag applies',
+                                 on_delete=models.CASCADE)
+
+    class Meta:
+        unique_together = (('user', 'resource', 'kind'),)
```
Comment (pkdash) [inline hs_labels/models.py]: Since this a singe set of fields for uniqueness, it can bet set as a single tuple ```('user', 'resource', 'kind')```

File: hs_labels/models.py [hunk 5]
```diff
@@ -99,25 +112,37 @@ class UserResourceFlags(models.Model):
...
-    user = models.ForeignKey(User, null=True, help_text='user who stored the label', related_name='ul2usl')
+    Storage class for persistent labels that are reusable across different kinds of objects
+    """
+    user  = models.ForeignKey(User, null=False,
+                              help_text='user who stored the label',
+                              related_name='ul2usl',
+                              on_delete=models.CASCADE)
     label = models.TextField(help_text='label to be stored by user')
+    class Meta:
+        unique_together = (('user', 'label'),)
```
Comment (pkdash) [inline hs_labels/models.py]: as above.

File: hs_labels/tests/test_resource_labelling.py [hunk 6]
```diff
@@ -105,94 +120,94 @@ def test_labels(self):
         cat = self.cat  # user
         scratching = self.scratching  # resource
         cat.ulabels.label_resource(scratching, "penalty clause")
```
Comment (pkdash) [inline hs_labels/tests/test_resource_labelling.py]: Should include test to show that duplicate labels can't be assigned to a resource (duplicate record problem).

File: hs_labels/models.py [hunk 7]
```diff
@@ -199,113 +233,157 @@ def clean_label(name):
...
+        if __debug__:  # during testing only, check argument types and preconditions
+            assert isinstance(this_resource, BaseResource)
+            assert isinstance(this_label, basestring)
 
         # remove leading and trailing spaces
         label_string = UserLabels.clean_label(this_label)
-        UserResourceLabels.objects.get_or_create(resource=this_resource,
-                                                 label=label_string,
-                                                 user=self.user)
+        with transaction.atomic():  # empirically, get_or_create is not atomic.
```
Comment (pkdash) [inline hs_labels/models.py]: Is there unit test to show that using ```transaction.atomic()``` serves the purpose? I am not sure what benefits we get using ```transaction.atomic()``` here.

File: hs_labels/models.py [hunk 8]
```diff
@@ -199,113 +233,157 @@ def clean_label(name):
...
+        Users are allowed to flag any resource, including resources to which they do not have access.
+        This is not an access control problem because flagging information is private.
+        """
+        if __debug__:  # during testing only, check argument types and preconditions
+            assert isinstance(this_resource, BaseResource)
+            assert this_flagcode == FlagCodes.FAVORITE or this_flagcode == FlagCodes.MINE
 
-        Users are allowed to file any resource, including resources to which they do not have access.
-        This is not an access control problem because labeling information is private.
+        with transaction.atomic():  # empirically, get_or_create is not atomic.
```
Comment (pkdash) [inline hs_labels/models.py]: as above.

File: hs_labels/models.py [hunk 9]
```diff
@@ -199,113 +233,157 @@ def clean_label(name):
...
+            assert isinstance(this_label, basestring)
+        UserResourceLabels.objects.filter(label=this_label, user=self.user)\
+                                  .delete()
 
-    def favorite_resource(self, this_resource):
+    ##########################################
+    # general flagging of resources
+    ##########################################
+
+    def flag_resource(self, this_resource, this_flagcode):
```
Comment (pkdash) [inline hs_labels/models.py]: Should this helper function be private? Otherwise, this should have corresponding unit tests.

File: hs_labels/models.py [hunk 10]
```diff
@@ -199,113 +233,157 @@ def clean_label(name):
...
 
-        Users are allowed to file any resource, including resources to which they do not have access.
-        This is not an access control problem because labeling information is private.
+        with transaction.atomic():  # empirically, get_or_create is not atomic.
+            UserResourceFlags.objects.get_or_create(resource=this_resource,
+                                                     kind=this_flagcode,
+                                                     user=self.user)
 
-        Resource gets flagged as favorite if it is not already flagged as favorite
+    def unflag_resource(self, this_resource, this_flagcode):
```
Comment (pkdash) [inline hs_labels/models.py]: same as above.

File: hs_labels/models.py [hunk 11]
```diff
@@ -402,22 +463,27 @@ def get_labels(self, this_user):
                                    .order_by("label").all()
         return labels
 
-    def is_favorite(self, this_user):
-        """ return True if this resource has been flagged as favorite by a given user
-
-        :return: True or False
+    def is_flagged(self, this_user, this_flagcode):
```
Comment (pkdash) [inline hs_labels/models.py]: same as above.

File: hs_labels/migrations/0003_auto_20160107_1847.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
```
Comment (pkdash) [inline hs_labels/migrations/0003_auto_20160107_1847.py]: Consider changing the migration file name (e.g, 0003_custom_migration_20160107_1847.py) as the current name implies it was automatically generated by Django.

File: hs_labels/models.py (hunk 7 above)
Comment (alvacouch) [inline hs_labels/models.py]: Hellishly difficult. If we can't trust transaction.atomic(), we can't trust anything!

File: hs_labels/models.py (hunk 9 above)
Comment (alvacouch) [inline hs_labels/models.py]: Will write unit tests and would like to deprecate the non-helper functions. This seems a better interface overall.

**Comment by alvacouch** [comment]
I have made changes as indicated except as follows: 
*  I do not consider testing that transactions work to be practical at this time. I suggest trusting that they work.

File: hs_labels/models.py (hunk 9 above)
Comment (alvacouch) [inline hs_labels/models.py]: Unit tests written.

File: hs_labels/migrations/0003_manual_delete_duplicates.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+
+def delete_duplicates(apps, schema_editor): 
+    # delete oldest copies of records from UserResourceLabels
+    UserResourceLabels = apps.get_model("hs_labels", "UserResourceLabels")
+    oldrecord = None 
+    # order by reverse start so that the first object should be preserved
+    for record in list(UserResourceLabels.objects.all()\
```
Comment (pkdash) [inline hs_labels/migrations/0003_manual_delete_duplicates.py]: Does this need to be converted to a list?

File: hs_labels/migrations/0003_manual_delete_duplicates.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+	and oldrecord.label==record.label): 
+	    record.delete() 
+	else: 
+	    oldrecord = record
+
+    # delete oldest copies of records from UserResourceFlags
+    UserResourceFlags = apps.get_model("hs_labels", "UserResourceFlags")
+    oldrecord = None 
+    # order by reverse start so that the first object should be preserved
+    for record in UserResourceFlags.objects.all()\
```
Comment (pkdash) [inline hs_labels/migrations/0003_manual_delete_duplicates.py]: If the above one needs to be converted to a list then this one also should be.

File: hs_labels/migrations/0003_manual_delete_duplicates.py (hunk 13 above)
Comment (alvacouch) [inline hs_labels/migrations/0003_manual_delete_duplicates.py]: Yes, because we are deleting rows. The default is to iterate abstractly over the QuerySet. We wish to avoid that because we are deleting things and that has some small chance of screwing up iteration.

**Comment by pkdash** [comment]
+1
